### PR TITLE
Added  type guard overload ReadonlyArray.prototype.every

### LINKED ISSUE #14963: Type guards in Array.prototype.every
**TypeScript Version:**  2.2.2

**Code**

```ts
function isNumber(x: any): x is number { return typeof(x) === "number"; }

let numbersOrStrings: (number | string)[] = [1, 2, 3, 4];

if(numbersOrStrings.every(isNumber)) {
    let _numbers: number[] = numbersOrStrings;
    // ...
}
```

**Expected behavior:**

When every member of `numberOfStrings` matches the `isNumber` type guard, the whole array should be inferred to be `numberOfStrings: number[]`; so, the above code should compile without error.

**Actual behavior:**

The above fails to compile, with the following error:

```
error TS2322: Type '(string | number)[]' is not assignable to type 'number[]'.
```

---

This issue is similar to #7657, and it can be solved by adding something like the following to `lib.d.ts`:

```ts
interface Array<T> {
    every<U extends T>(pred: (a: T) => a is U): this is U[];
}
```

(or, the above can be added locally to a project as a temporary workaround)

### PR DESCRIPTION
**Pull request by NaridaL**
Fixes #14963

### CODE REVIEW

**Comment by NaridaL** [comment]
Do the test.symbols and the test.types files not get generated if there are error? Is there a better way of testing for errors so that those still get generated?

File: tests/baselines/reference/arrayEvery.errors.txt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    }, "foo"); // first overload
+    baseReadonlyArray.every(function (x) {
+      return this.indexOf("f") > 0;
+    }, 1); // error: thisArg type doesn't match
+    
+    if (baseReadonlyArray.every<number>((x): x is number => "number" === typeof x)) {
+      const numberReadonlyArray: ReadonlyArray<number> = baseReadonlyArray; // should be ReadonlyArray<number>
+    }
+    
+    if (baseReadonlyArray.every((x): x is number => "number" === typeof x)) {
```
Comment (NaridaL) [inline tests/baselines/reference/arrayEvery.errors.txt]: OK, something's not working with the inference...?

**Comment by RyanCavanaugh** [comment]
> Do the test.symbols and the test.types files not get generated if there are errors?

Correct. Best practice is to have one testcase with all the error cases, and one testcase with zero errors. This is also better for future reviews since a file going from 0 to 1 errors will raise a red flag more quickly than a file going from e.g. 4 to 5.

**Comment by NaridaL** [comment]
OK, in the current state, there should be no errors. For some reason it's not working if the type parameter isn't specified explicitly. I'm not sure what the issue is.

**Comment by NaridaL** [comment]
OK, the issue was https://github.com/Microsoft/TypeScript/issues/18562, I've used a workaround, the tests pass as they should. This PR should be mergeable as it is now.

**Comment by NaridaL** [comment]
@RyanCavanaugh any comment on the current state of the PR?

**Comment by RyanCavanaugh** [comment]
@mhegazy 👍 ?

**Comment by mhegazy** [comment]
We should have a similar change in Array as well. Array and ReadonlyArray should stay in sync.

**Comment by NaridaL** [comment]
@mhegazy Does that make sense? according to the linked issue
```
declare const strNumArr: (string | number)[]
const strNumArr2: (string | number)[] = strNumArr
if (strNumArr.every(isString)) {
   strNumArr2.push(23)
   // strNumArr now has type string[] and contains a number
}
```

... although tbh you can get the same prob with a ReadonlyArray with only slightly more effort.

**Comment by NaridaL** [comment]
@mhegazy Considering my last comment, this should be mergeable as-is? Or do we want to revisit the original issue?

**Comment by shanehsu** [comment]
@NaridaL Hope this get merged soon. Relying on `.filter` now. 👍

**Comment by DanielRosenwasser** [comment]
I think we're just waiting on having the appropriate overload in `Array` for consistency

**Comment by NaridaL** [comment]
@mhegazy 

`.filter()` creates a new array. `.every()` doesn't so including the same overload for `Array` could lead to issues such as the one I describe here https://github.com/Microsoft/TypeScript/pull/18372#issuecomment-335382998. Does that not make sense? As far as I can tell this is the same conclusion reached in the base issue. #14963

**Comment by shanehsu** [comment]
This is a language issue then. TypeScript compiler have to track every reference assignment. Out of our hands now I suppose.

**Comment by simonbuchan** [comment]
@mhegazy I just hit this now, would you like an updated PR with the `Array` variant? Should I keep the current test location, or would it make sense to put it in `tests/cases/conformance/experssions/typeGuards`?

I noticed `tests/cases/compiler/arrayFilter.ts` doesn't test the type-guardiness.

**Pull request closed without merging**